File: Windows/src/NeoRemote.Core/src/MouseEventPlanner.cpp
#include "NeoRemote/Core/MouseEventPlanner.hpp"

#include <cmath>

namespace NeoRemote::Core {

PlannedMouseEvent PlannedMouseEvent::Move(Point pointValue)
{
    PlannedMouseEvent event;
    event.type = PlannedMouseEventType::Move;
    event.point = pointValue;
    return event;
}

PlannedMouseEvent PlannedMouseEvent::MouseDown(MouseButtonKind buttonValue, Point pointValue)
{
    PlannedMouseEvent event;
    event.type = PlannedMouseEventType::MouseDown;
    event.button = buttonValue;
    event.point = pointValue;
    return event;
}

PlannedMouseEvent PlannedMouseEvent::MouseUp(MouseButtonKind buttonValue, Point pointValue)
{
    PlannedMouseEvent event;
    event.type = PlannedMouseEventType::MouseUp;
    event.button = buttonValue;
    event.point = pointValue;
    return event;
}

PlannedMouseEvent PlannedMouseEvent::Drag(MouseButtonKind buttonValue, Point pointValue)
{
    PlannedMouseEvent event;
    event.type = PlannedMouseEventType::Drag;
    event.button = buttonValue;
    event.point = pointValue;
    return event;
}

PlannedMouseEvent PlannedMouseEvent::Scroll(int lines)
{
    PlannedMouseEvent event;
    event.type = PlannedMouseEventType::Scroll;
    event.scrollLines = lines;
    return event;
}

std::vector<PlannedMouseEvent> MouseEventPlanner::Apply(const RemoteCommand& command, const PositionProvider& positionProvider)
{
    if (!currentPosition_) {
        currentPosition_ = positionProvider();
    }

    switch (command.type) {
    case RemoteCommandType::Move:
        return {PlannedMouseEvent::Move(TranslatedPoint(command.dx, command.dy))};

    case RemoteCommandType::Tap: {
        const Point point = currentPosition_.value_or(positionProvider());
        return {
            PlannedMouseEvent::MouseDown(command.button, point),
            PlannedMouseEvent::MouseUp(command.button, point),
        };
    }

    case RemoteCommandType::Scroll:
        return {PlannedMouseEvent::Scroll(static_cast<int>(std::llround(command.deltaY)))};

    case RemoteCommandType::Drag:
        switch (command.state) {
        case DragState::Started: {
            const Point point = currentPosition_.value_or(positionProvider());
            activeDragButton_ = MouseButtonKind::Primary;
            return {PlannedMouseEvent::MouseDown(MouseButtonKind::Primary, point)};
        }
        case DragState::Changed: {
            const MouseButtonKind button = activeDragButton_.value_or(MouseButtonKind::Primary);
            activeDragButton_ = button;
            return {PlannedMouseEvent::Drag(button, TranslatedPoint(command.dx, command.dy))};
        }
        case DragState::Ended: {
            const Point point = currentPosition_.value_or(positionProvider());
            const MouseButtonKind button = activeDragButton_.value_or(MouseButtonKind::Primary);
            activeDragButton_.reset();
            return {PlannedMouseEvent::MouseUp(button, point)};
        }
        }
        break;

    case RemoteCommandType::Heartbeat:
        return {};
    }

    return {};
}

std::optional<Point> MouseEventPlanner::CurrentPosition() const
{
    return currentPosition_;
}

Point MouseEventPlanner::TranslatedPoint(double dx, double dy)
{
    Point current = currentPosition_.value_or(Point{});
    current.x += dx;
    current.y += dy;
    currentPosition_ = current;
    return current;
}

} // namespace NeoRemote::Core

File: MacOS/Sources/NeoRemoteMac/Core/InputInjection/MouseEventPlanner.swift
import CoreGraphics
import Foundation

enum PlannedMouseEvent: Equatable {
    case move(to: CGPoint)
    case mouseDown(button: MouseButtonKind, point: CGPoint)
    case mouseUp(button: MouseButtonKind, point: CGPoint)
    case drag(button: MouseButtonKind, to: CGPoint)
    case scroll(deltaX: Int32, deltaY: Int32)
}

struct MouseEventPlanner {
    private(set) var currentPosition: CGPoint?
    private(set) var activeDragButton: MouseButtonKind?

    mutating func apply(_ command: RemoteCommand, positionProvider: () -> CGPoint) -> [PlannedMouseEvent] {
        if currentPosition == nil {
            currentPosition = positionProvider()
        }

        switch command {
        case .clientHello:
            return []

        case let .move(dx, dy):
            let point = translatedPoint(dx: dx, dy: dy)
            return [.move(to: point)]

        case let .tap(button):
            let point = currentPosition ?? positionProvider()
            return [
                .mouseDown(button: button, point: point),
                .mouseUp(button: button, point: point),
            ]

        case let .scroll(deltaX, deltaY):
            return [
                .scroll(
                    deltaX: Int32(deltaX.rounded()),
                    deltaY: Int32(deltaY.rounded())
                ),
            ]

        case let .drag(state, button, dx, dy):
            switch state {
            case .started:
                let point = currentPosition ?? positionProvider()
                activeDragButton = button
                return [.mouseDown(button: button, point: point)]

            case .changed:
                let point = translatedPoint(dx: dx, dy: dy)
                let button = activeDragButton ?? .primary
                activeDragButton = button
                return [.drag(button: button, to: point)]

            case .ended:
                let point = currentPosition ?? positionProvider()
                let button = activeDragButton ?? .primary
                activeDragButton = nil
                return [.mouseUp(button: button, point: point)]
            }

        case .heartbeat:
            return []
        }
    }

    private mutating func translatedPoint(dx: Double, dy: Double) -> CGPoint {
        let current = currentPosition ?? .zero
        let point = CGPoint(x: current.x + dx, y: current.y + dy)
        currentPosition = point
        return point
    }
}
